SCC-2207 Add holding class for retrieval

mwbenowitz opened the pull request:
To integrate Holdings data into the discovery API, we need to fetch the statements for the holdings separate from the item statements.

The query SQL query fetches simply an array of statements associated with a bib (or multiple bibs) record at uses this to construct the item objects. I think it would be possible to modify that query (possibly with a `FILTER` in the `item_statement` aggregation), but I think that might be difficult to maintain and understand in the future. This simply checks the identifier prefix for each statement returned in the `item_statements` array and sorts them on if it is an `i` or `h`. This is a simple approach that I don't think poses any performance issues since it is such a simple check.

Some questions I have on this approach:

- Should we rename the `item_statement` in the query? I've left it as is for now since I couldn't think of anything better
- Is there any potential for side effects here? I've tested the code and I believe it will just pass through item statements if no holding statements are present.
- Philosophically, is this better to be done in the query? I could see an argument for that, but I actually worry it would have a greater performance impact since it would involve adding another subquery, though in likelihood, that would have less of an impact that running `filter` on the existing array.

mwbenowitz committed:
SCC-2207 Add holding class for retrieval

To integrate Holdings data into the discovery API, we need to fetch the statements for the holdings separate from the item statements.

The query SQL query fetches simply an array of statements associated with a bib (or multiple bibs) record at uses this to construct the item objects. I think it would be possible to modify that query (possibly with a `FILTER` in the `item_statement` aggregation), but I think that might be difficult to maintain and understand in the future. This simply checks the identifier prefix for each statement returned in the `item_statements` array and sorts them on if it is an `i` or `h`. This is a simple approach that I don't think poses any performance issues since it is such a simple check.

Some questions I have on this approach:

- Should we rename the `item_statement` in the query? I've left it as is for now since I couldn't think of anything better
- Is there any potential for side effects here? I've tested the code and I believe it will just pass through item statements if no holding statements are present.
- Philosophically, is this better to be done in the query? I could see an argument for that, but I actually worry it would have a greater performance impact since it would involve adding another subquery, though in likelihood, that would have less of an impact that running `filter` on the existing array.

nonword commented on lib/db.js:
```diff
@@ -133,7 +160,8 @@ function _bibSqlQueryString (options) {
     (
-      SELECT ${_jsonAggSelectString('_IS', { additionalColumns: { s: 'subject_id', bn: 'blanknode' } })} AS statements
+      SELECT 
+        ${_jsonAggSelectString('_IS', { additionalColumns: { s: 'subject_id', bn: 'blanknode' } })} FILTER (WHERE AS statements
```
Is that `FILTER (WHERE` supposed to be there?

mwbenowitz replied:
thanks for catching that! I was considering implementing the filter in the query when I realized that it could be done outside of the SQL and switched approaches, but forgot to remove this (of course the one time I didn't git diff before adding a file is when something like this is in there)

nonword reviewed (commented):
Looks good except for comment some rogue sql. The approach seems sensible to me. I think I agree that an in-sql segmentation would be preferable, but an in-code segmentation is less risky and probably easier to maintain and reason about. As a mechanism, segmenting on prefix seems fine. Another option would be grouping on rdfs:type, but that would require first gathering up all statements by "uri", so there'd be some overhead. I think prefix is good.

mwbenowitz committed:
SCC-2207 Remove dangling SQL fragment

Stray SQL fragment from an earlier attempt at implementing this change

mwbenowitz committed:
SCC-2207 Update bib to include holdings statements

This previously omitted the necessary step of adding the `_holdings` array to the output bib object, and this includes it. It is generated in the same way that the `_items` array is updated.

This also bumps the version to `1.3.0` as this should probably be considered a minor version update

mwbenowitz committed:
SCC-2207 Update bib to include holdings statements

This previously omitted the necessary step of adding the `_holdings` array to the output bib object, and this includes it. It is generated in the same way that the `_items` array is updated.

This also bumps the version to `1.3.0` as this should probably be considered a minor version update

mwbenowitz commented:
@nonword I realized that I had actually forgot to add the holding statements to the returned `Bib` record, so just added that in.

And I think this should be released as `1.3.0` It's not a breaking change, but it is adding some significant functionality.

Thanks for your help with all these PRs!

mwbenowitz committed:
Merge branch 'SCC-2207-add-holding-class' of https://github.com/NYPL-discovery/discovery-store-models into SCC-2207-add-holding-class

mwbenowitz committed:
SCC-2207 Additional bugfix

In sorting the item from holding statements, the item statements were being sorted first, which removed all the holding statements before they could be filtered into their own array. Inverting the order of these filter calls resolves the problem.

mwbenowitz committed:
SCC-2207 Fix handling of blanknodes on holdings

Previously this omitted the data from blanknodes on holding records. This borrows a pattern from the `Bib` model to represent these nodes.

mwbenowitz committed:
SCC-2207 Add test coverage for holdings
